### PR TITLE
[mlir][Vector] Fix vector.insert folder for scalar to 0-d inserts

### PR DESCRIPTION
**Pull request by Groverkss**
The current vector.insert folder tries to replace a scalar with a 0-rank vector. This patch fixes this crash by not folding unless they types of the result and replacement are same.

### CODE REVIEW

File: mlir/test/Dialect/Vector/canonicalize.mlir
Comment (banach-space) [inline mlir/test/Dialect/Vector/canonicalize.mlir:2756]: Could this be added near other tests for `vector::InsertOp::fold` so that all cases are clustered together? Then it will be much easier to see what's already tested and what isn't. And what cases are supported. Thanks!

Also,  I'd avoid "regression" in test function names and instead use something that helps understand what the exercises corner case is - currently that's not clear.  How about: `@insert_into_0d_ regression` -> `@insert_no_fold_type_mismatch`.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2958]:
```suggestion
  if (getNumIndices() == 0 && getSourceType() == getType())
```

File: mlir/test/Dialect/Vector/canonicalize.mlir
Comment (kuhar) [reply mlir/test/Dialect/Vector/canonicalize.mlir:2756]: +1

Comment (Groverkss) [reply mlir/test/Dialect/Vector/canonicalize.mlir:2756]: Sure, I will change the name. The reason for the name was, ExtractOp::fold had the same bug before and extract_from_0d_regression is the test name for that :)

Comment (Groverkss) [reply mlir/test/Dialect/Vector/canonicalize.mlir:2756]: There are no tests for vector::InsertOp::Fold. So i'm not sure where to cluster them. The place i put this test, is the closest to a relevant vector.insert fold pattern.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp
Comment (Groverkss) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2958]: I'll change it, but that `getType()` looks more confusing than `getResultType().getType()`, unless you know what the method defines as `getType()`.

File: mlir/test/Dialect/Vector/canonicalize.mlir
Comment (banach-space) [reply mlir/test/Dialect/Vector/canonicalize.mlir]: > The reason for the name was, ExtractOp::fold had the same bug before and extract_from_0d_regression is the test name for that :)

Consistency is key, I agree. Let's rename `@extract_from_0d_regression` as well so that tests for `vector.insert` and `vector.extract` are symmetrical (at least for the logic that's nearly identical).

> There are no tests for vector::InsertOp::Fold. 

This explains why we keep discovering these bugs - if we don't test the folder, we can't really claim it "works" ™️ . Let's fix that! 

Could you please add tests for `vector::InsertOp::Fold`? Both positive (missing) and negative (already included in this PR). Some Ops follow this naming convention that makes sense to me:
* `@<op_name>_fold{_<suffix>}` - positive tests,
* `@<op_name>_no_fold{_<suffix>}` - negative tests.

Sadly, tests for `vector.extract` are more like `@fold_extract_shapecast` 😢 Pick one style and I'll try to unify things when adding tests for scalable vectors.

You can add the new tests near where `ExtractOp::fold` is tested (e.g. after `@dont_fold_0d_extract_shapecast`). I can't think of a better place just now.

Thanks for improving this!

Comment (Groverkss) [reply mlir/test/Dialect/Vector/canonicalize.mlir]: - Renamed extract_from_0d_regression -> extract_no_fold_scalar_to_0d
- Added insert folder tests: insert_fold_same_rank, insert_no_fold_scalar_to_0d
- Moved insert tests right after extract folding tests

Let me know if i missed something :)

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

Please wait for +1 from @kuhar before landing this.

**Review by hanhanW — APPROVED** [review]

**Review by dcaballe — APPROVED** [review]
Thanks for the fix!

**Review by kuhar — APPROVED** [review]

**Pull request merged**
